src/java/org/apache/cassandra/index/sai/disk/v1/SegmentBuilder.java

```diff
@@ -248,12 +283,56 @@ public long add(ByteBuffer term, PrimaryKey key, long sstableRowId)
 
         maxSegmentRowId = Math.max(maxSegmentRowId, segmentRowId);
 
-        long bytesAllocated = addInternal(term, segmentRowId);
+        long bytesAllocated = supportsAsyncAdd()
+                              ? addInternalAsync(term, segmentRowId)
+                              : addInternal(term, segmentRowId);
         totalBytesAllocated += bytesAllocated;
 
         return bytesAllocated;
     }
 
+    private long addInternalAsync(ByteBuffer term, int segmentRowId)
+    {
+        updatesInFlight.incrementAndGet();
+        ForkJoinPool.commonPool().submit(() -> {
+            try
+            {
+                long bytesAdded = addInternal(term, segmentRowId);
+                termSizeReservoir.update(bytesAdded);
```

Comment:
Perhaps we can have a counter that also keeps track of the real size so our logging on completion is accurate? That might help us understand if there are any issues with excessively large indexes, too.

Reply:
I left that out b/c the vector terms will be exactly uniform, but it's good future proofing if and when someone turns it on for other data types.  Will add it.